### PR TITLE
Restructure of 'ck_checkpoints'

### PR DESCRIPTION
**Pull request by shipyy**

### CODE REVIEW

**Comment by tslashd** [comment]
1. Test DB
2. [Decimal Conversion](https://github.com/surftimer/SurfTimer/blob/master/scripts/upgrade_scripts/decimalruntimes.sql)
3. [Checkpoints Conversion](https://github.com/shipyy/Surftimer-Official/blob/ck-checkpoints-restructure/scripts/upgrade_scripts/upgrade_ck_checkpoints.sql)
4. Server starts fine
5. Someone joins server

```
L 07/08/2022 - 19:53:51: [SM] Exception reported: Nope. Line: 184, Error: Lost connection to MySQL server during query
L 07/08/2022 - 19:53:51: [SM] Blaming: SurfTimer.smx
L 07/08/2022 - 19:53:51: [SM] Call stack trace:
L 07/08/2022 - 19:53:51: [SM]   [0] SetFailState
L 07/08/2022 - 19:53:51: [SM]   [1] Line 184, surftimer\db\updater.sp::SQLCheckDataType
```

**Review by Bara — CHANGES_REQUESTED** [review]

File: addons/sourcemod/scripting/surftimer/db/updater.sp [hunk 1]
```diff
@@ -130,6 +134,26 @@ public void db_upgradeDatabase(int ver)
 	{
 		SQL_FastQuery(g_hDb, sql_createReplays);
 	}
+	else if (ver == 9)
+	{
+		char sql_checkDataTyped[512];
+		Format(sql_checkDataTyped, sizeof(sql_checkDataTyped), "SELECT DATA_TYPE, NUMERIC_PRECISION, NUMERIC_SCALE FROM information_schema.COLUMNS WHERE TABLE_SCHEMA='%s' AND TABLE_NAME='ck_checkpoints' AND COLUMN_NAME='cp1' HAVING DATA_TYPE = 'decimal' AND NUMERIC_PRECISION = 12 AND NUMERIC_SCALE = 6",g_sDatabaseName);
```
[the remark this answers is not in the document]
Comment (shipyy) [reply addons/sourcemod/scripting/surftimer/db/updater.sp:140]: u mean somethjing like this?
image

or am i complicating? xD

File: addons/sourcemod/scripting/surftimer/db/updater.sp [hunk 2]
```diff
@@ -138,35 +162,18 @@ void LoopFloatDecimalTables()
...
-	else
-	{
-		strcopy(sColumn, sizeof(sColumn), column);
-	}
+	strcopy(sColumn, sizeof(sColumn), column);
 
 	char sQuery[512];
 	Format(sQuery, sizeof(sQuery), "SELECT TABLE_NAME, COLUMN_NAME, DATA_TYPE, NUMERIC_PRECISION, NUMERIC_SCALE FROM information_schema.COLUMNS WHERE TABLE_SCHEMA='%s' AND TABLE_NAME='%s' AND COLUMN_NAME='%s';", g_sDatabaseName, table, sColumn);
-
+	PrintToServer(sQuery);
```
[the remark this answers is not in the document]
Comment (shipyy) [reply addons/sourcemod/scripting/surftimer/db/updater.sp:176]: yup

File: addons/sourcemod/scripting/surftimer/sql.sp [hunk 3]
```diff
@@ -3499,17 +3502,28 @@ public void db_UpdateCheckpoints(int client, char szSteamID[32], int zGroup)
...
-		SQL_TQuery(g_hDb, SQL_updateCheckpointsCallback, szQuery, pack, DBPrio_Low);
+		for(int i = 0; i < cp_count; i++){
+			Format(szQuery, sizeof(szQuery), sql_updateCheckpoints, g_fCheckpointTimesNew[zGroup][client][i], szSteamID, g_szMapName, zGroup);
+			SQL_TQuery(g_hDb, SQL_updateCheckpointsCallback, szQuery, pack, DBPrio_Low);
+		}
```
Comment (coverrt) [inline addons/sourcemod/scripting/surftimer/sql.sp:3518]: Transactions should be implemented here, correct?

File: addons/sourcemod/scripting/surftimer/sql.sp [hunk 4]
```diff
@@ -3499,17 +3502,28 @@ public void db_UpdateCheckpoints(int client, char szSteamID[32], int zGroup)
...
-		SQL_TQuery(g_hDb, SQL_updateCheckpointsCallback, szQuery, pack, DBPrio_Low);
+		for(int i = 0; i < cp_count; i++){
+			Format(szQuery, sizeof(szQuery), sql_insertCheckpoints, szSteamID, g_szMapName, g_fCheckpointTimesNew[zGroup][client][i], zGroup);
+			SQL_TQuery(g_hDb, SQL_updateCheckpointsCallback, szQuery, pack, DBPrio_Low);
+		}
```
Comment (coverrt) [inline addons/sourcemod/scripting/surftimer/sql.sp:3526]: Transactions should be implemented here, correct?

File: addons/sourcemod/scripting/surftimer/sql.sp (hunk 3 above)
Comment (shipyy) [reply addons/sourcemod/scripting/surftimer/sql.sp:3518]: what u mean? i am confused xD

File: addons/sourcemod/scripting/surftimer/db/updater.sp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Bara) [reply addons/sourcemod/scripting/surftimer/db/updater.sp]: I missed that SQL_FastQuery didn't contain `any ...`... but [this](https://github.com/surftimer/SurfTimer/pull/472/commits/a106047f582fc39034b19af32f844e6452507803) looks similar to the other queries.sp stuff

File: addons/sourcemod/scripting/surftimer/sql.sp (hunk 3 above)
Comment (Bara) [reply addons/sourcemod/scripting/surftimer/sql.sp:3518]: yes @coverrt would be better with 32 queries at the same time

**Review by Bara — APPROVED** [review]
Would be nice if someone could test this again ( @tslashd maybe? )

**Comment by tslashd** [comment]
Conversion was successful
:PagMan:

**Pull request merged**